feat: removing visible_date-to-creds updates per-cert

deborahgu opened the pull request:
## Description

The credentials IDA now relies on  the course certificate configuration and (if present) `certificate_available_date` for displayability. We no longer need to send `visible_date` updates for every awarded certificate when a course  overview changes.

FIXES: APER-3535

deborahgu committed:
feat: removing visible_date-to-creds updates per-cert

The credentials IDA now relies on  the course certificate configuration
and (if present) `certificate_available_date` for displayability. We no
longer need to send `visible_date` updates for every awarded certificate
when a course  overview changes.

FIXES: APER-3535

deborahgu committed:
Merge branch 'master' into dkaplan1/APER-3535_stop-calculating-and-sending-a-visible-date-to-credentials

deborahgu committed:
feat: removing visible_date-to-creds updates per-cert

Modifying the tests to reflect that we will still send date_override but
not visible_date on `POST`.

FIXES: APER-3535

deborahgu commented on cms/djangoapps/contentstore/docs/diagrams/certificate_available_date.dsl:
```diff
@@ -1,6 +1,6 @@
 /*
- * This is a high level diagram visualizing how the `CERTIFICATE_AVAILBLE_DATE` and "visible date" attribute updates
```
can I just state how much easier it made this documentation update that image was generated with DSL  input which was also next to it in the repository? 🙏 @justinhynes

deborahgu commented on lms/djangoapps/certificates/api.py:
```diff
@@ -849,7 +849,7 @@ def _course_uses_available_date(course):
     )
 
 
-def available_date_for_certificate(course, certificate):
```
I added a fair amount of type checking which you will see  here and below, because it became a side effect of my understanding what the methods did so I  could understand them enough to safely modify or remove.

deborahgu commented on openedx/core/djangoapps/catalog/utils.py:
```diff
@@ -31,6 +32,9 @@
 from openedx.core.djangoapps.oauth_dispatch.jwt import create_jwt_for_user
 from openedx.core.lib.edx_api_utils import get_api_data
 
+if TYPE_CHECKING:
```
anything in a `if TYPE_CHECKING` block isn't actually imported, it's just imported during linter/mypy type checks.

deborahgu commented on openedx/core/djangoapps/catalog/utils.py:
```diff
@@ -98,7 +102,14 @@ def check_catalog_integration_and_get_user(error_message_field):
 # pylint: disable=redefined-outer-name
-def get_programs(site=None, uuid=None, uuids=None, course=None, catalog_course_uuid=None, organization=None):
+def get_programs(
+    site: "Site" = None,
```
if a type is only imported for type checks, putting it in quotation marks, as here, will allow the type checker to do its thing without actually causing an import error during runtime.

deborahgu commented on openedx/core/djangoapps/catalog/utils.py:
```diff
@@ -112,7 +123,7 @@ def get_programs(site=None, uuid=None, uuids=None, course=None, catalog_course_u
 
     Returns:
-        list of dict, representing programs.
+        list of str, representing programs.
```
comment change just reflects what I believe was the state of affairs. I do not believe that this returns a list of dicts.

deborahgu commented on openedx/core/djangoapps/programs/signals.py:
```diff
@@ -2,7 +2,6 @@
 This module contains signals / handlers related to programs.
```
This is almost entirely linter improvements that I missed on my first past of lint-only changes, except for the 4 lines removed at the bottom.

deborahgu commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -53,14 +59,14 @@ def get_completed_programs(site, student):
     """
     meter = ProgramProgressMeter(site, student)
-    return meter.completed_programs_with_available_dates
+    return set(meter.completed_programs_with_available_dates.keys())
```
after talking with @justinhynes, instead of modifying `completed_programs_with_available_dates`, I'm doing the safer thing and just taking the current return value and extracting out just the set of program UUIDs.  Re-factoring unnecessary code out of `ProgramProgressMeter` would be lovely but there are larger infrastructural improvements that would make that easier.

justinhynes replied:
I'm a lil worried about this change actually. I'm worried it might not work as intended. 

It looks like this piece of code is only used [right now in the `award_program_certificates` task](https://github.com/openedx/edx-platform/blob/e02a22f33847ba17db84cd7549fe4fc93c54f5fb/openedx/core/djangoapps/programs/tasks.py#L311-L314):
```
    try:
        completed_programs = {}
        for site in Site.objects.all():
            completed_programs.update(get_completed_programs(site, student))
```
We changed the return type of `get_completed_programs()` to a set, and `completed_programs` is a dict. I ran these changes locally in devstack after I got worried and I'm encountering a ValueError exception when we try to update the `completed_programs.update()` dict: `ValueError('dictionary update sequence element #0 has length 36; 2 is required')`

deborahgu committed:
feat: removing visible_date-to-creds updates per-cert

Fixing a linter directive.

FIXES: APER-3535

deborahgu committed:
feat: removing visible_date-to-creds updates per-cert

Removing a missed test for visible_date attributes

FIXES: APER-3535

justinhynes commented on cms/djangoapps/contentstore/docs/diagrams/certificate_available_date.dsl:
```diff
@@ -1,6 +1,6 @@
- * This is a high level diagram visualizing how the `CERTIFICATE_AVAILBLE_DATE` and "visible date" attribute updates
- * are updated internally and transmit to the Credentials IDA.
+ * This is a high level diagram visualizing how the `CERTIFICATE_AVAILBLE_DATE` update is
+ * are updated internally and transmitted to the Credentials IDA.
```
Nit: maybe nix the `are` here now?

justinhynes commented on openedx/core/djangoapps/programs/docs/decisions/0001-sync-certificate-available-dates.rst:
```diff
@@ -4,7 +4,9 @@ Sync certificate_available_date and visible_date for certificates
 ------
 
-Review
+Superseded by Credentials ADR `0001 Certificate Available Date`_.
+
+.. _0001 Certificate Available Date: https://github.com/openedx/edx-platform/blob/master/openedx/core/djangoapps/oauth_dispatch/docs/decisions/0005-restricted-application-for-SSO.rst
```
Thank you for updating the ADR to point to the current state of the world.

deborahgu committed:
feat: removing visible_date-to-creds updates per-cert

Suggestion from code review on comment

FIXES: APER-3535

justinhynes commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -421,9 +424,8 @@ def update_credentials_course_certificate_configuration_available_date(
-    date will always either be the available date that is set in Studio for a
-    given course, or it will be None.
+    in Credentials. This date will always either be the available date that is
+    set in Studio for a given course, or it will be None.
```
This comment isn't entirely true anymore, and I missed updating it in an earlier round of updates (when we starting sending the end date as a CAD for courses with a CDB of "end").

The date will be, depending on the display behavior settings, either:
1) the end date of the course, or
2) the available date set in course settings of Studio, or
3) None. 

This is documented verbosely in a few other places though... so not sure we need all that detail... but anyway, what do you think?

deborahgu replied:
I'd argue the comment on the logic should be in the place the logic is defined, so I am cutting that final sentence.

justinhynes reviewed (changes requested):
Please see my comment w.r.t. the changes to the `get_completed_programs` function.

deborahgu committed:
Merge remote-tracking branch 'origin' into dkaplan1/APER-3535_stop-calculating-and-sending-a-visible-date-to-credentials

deborahgu committed:
feat: fixing a test

a merge conflict reintroducd this variable

deborahgu committed:
feat: fixing a mis-optimized return value from a utility method

code review pointed out that a modification to this method's return was
breaking the logic, but it hadn't been caught by any of the tests, so
the method was fixed but also a new test was added.

FIXES: APER-3535

deborahgu committed:
Merge branch 'master' into dkaplan1/APER-3535_stop-calculating-and-sending-a-visible-date-to-credentials

deborahgu committed:
feat: linter fix

unused import

FIXES: APER-3535

justinhynes commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -551,21 +552,21 @@ def award_course_certificate(self, username, course_run_key):
         f"Task award_course_certificate will award a course certificate to user {user.id} in course run "
-        f"{course_key} with a visible date of {visible_date}"
+        f"{course_key} with an available date of {available_date}"
     )
```
Is there any reason we need to calculate the `available_date` here or log it? If this is all handled by Credentials now, this seems unnecessary now?

deborahgu replied:
good catch, back to you

justinhynes commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -799,7 +748,7 @@ def update_certificate_available_date_on_course_update(self, course_key):
     # retry instead of failing it.
     if not is_credentials_enabled():
         error_msg = (
-            "Cannot execute the `update_certificate_visible_date_on_course_update` task. Issuing user credentials "
```
yikes. thanks for fixing this bad copy pasta 😅 (d'oh!)

deborahgu committed:
feat:  removing one unnecessary date calculation

one more unnecessary date calculation removed, as well as its associated
logging

FIXES: APER-3535

deborahgu committed:
Merge branch 'master' into dkaplan1/APER-3535_stop-calculating-and-sending-a-visible-date-to-credentials

deborahgu committed:
feat: Remove an import

removing an unnecessary import

justinhynes reviewed (approved):
This looks good! Thanks so much :D
